### ISSUE TITLE
Time between successive retries in Fluent Bit?

### ISSUE THREAD

**Issue by ujjwalgarg1995** [fluent/fluent-bit#3597]
There is a property called Retry_Limit but how to set time between successive retries?

```
[OUTPUT]
  Name            es
  Host            192.168.5.20
  Port            9200
  Logstash_Format On
  Retry_Limit     5
```

**Comment by JeffLuoo** [comment]
No option is available to configure that. However, the algorithm being used here is exponential backoff and jitter so that you can do some basic calculation to get a rough time you want: https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/

**Comment by donbowman** [comment]
Its not clear that its well defined. Here's an example from my log. I don't think the ~2000s backof is a good idea. Also, i don't think the algorithm as outlined is what its doing, empirically.

Also, it jumps from small to large and back. these are successive log messages.

In my case the retry come from another issue (the conflct detection in elastic), which in turn come from another issue (we retry the entire blob to elastic even if some passed).

```
[2021/06/08 12:45:27] [ warn] [engine] failed to flush chunk '1-1623148683.351335956.flb', retry in 1458 seconds: task_id=12, input=forward.1 > output=es.4 (out_id=4)
[2021/06/08 12:45:50] [ warn] [engine] failed to flush chunk '1-1623156349.894280401.flb', retry in 11 seconds: task_id=13, input=forward.1 > output=es.4 (out_id=4)
[2021/06/08 12:46:27] [ warn] [engine] failed to flush chunk '1-1623134394.911326215.flb', retry in 875 seconds: task_id=8, input=forward.1 > output=es.4 (out_id=4)
[2021/06/08 12:46:50] [ warn] [engine] failed to flush chunk '1-1623156409.894347586.flb', retry in 6 seconds: task_id=13, input=forward.1 > output=es.4 (out_id=4)
[2021/06/08 12:46:53] [ warn] [engine] failed to flush chunk '1-1623138933.358539142.flb', retry in 1891 seconds: task_id=9, input=forward.1 > output=es.4 (out_id=4)
[2021/06/08 12:47:10] [ warn] [engine] failed to flush chunk '1-1623156429.895968747.flb', retry in 6 seconds: task_id=13, input=forward.1 > output=es.4 (out_id=4)
[2021/06/08 12:47:30] [ warn] [engine] failed to flush chunk '1-1623156449.897197246.flb', retry in 9 seconds: task_id=13, input=forward.1 > output=es.4 (out_id=4)
[2021/06/08 12:47:50] [ warn] [engine] failed to flush chunk '1-1623156469.895645236.flb', retry in 11 seconds: task_id=13, input=forward.1 > output=es.4 (out_id=4)
[2021/06/08 12:48:18] [ warn] [engine] failed to flush chunk '1-1623112578.353349660.flb', retry in 180 seconds: task_id=3, input=forward.1 > output=es.4 (out_id=4)
[2021/06/08 12:48:50] [ warn] [engine] failed to flush chunk '1-1623156529.894393562.flb', retry in 10 seconds: task_id=13, input=forward.1 > output=es.4 (out_id=4)
```

**Comment by donbowman** [comment]
```
retry  0 -    4 backoff
retry  1 -    8 backoff
retry  2 -   25 backoff
retry  3 -  566 backoff
retry  4 - 1280 backoff
retry  5 - 1826 backoff
retry  6 - 1332 backoff
retry  7 -  933 backoff
retry  8 - 1598 backoff
retry  9 -  694 backoff
```


are the retries we generate usign the existing algorithm.

```
$ ./a.out 
retry  0 -    5 backoff
retry  1 -    8 backoff
retry  2 -   54 backoff
retry  3 -  998 backoff
retry  4 - 1016 backoff
retry  5 -  599 backoff
retry  6 -   96 backoff
retry  7 - 1379 backoff
retry  8 -  293 backoff
retry  9 -   12 backoff
```

w/ a lot of variation.


```
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <fcntl.h>
#include <time.h>

#define FLB_SCHED_CAP            2000
#define FLB_SCHED_BASE           5

static double ipow(double base, int exp)
{
    double result = 1;

    for (;;) {
        if (exp & 1) {
            result *= base;
        }

        exp >>= 1;
        if (!exp) {
            break;
        }
        base *= base;
    }

    return result;
}

int flb_random_bytes(unsigned char *buf, int len)
{
    int fd;
    int bytes;

    fd = open("/dev/urandom", O_RDONLY);
    if (fd == -1) {
        return -1;
    }

    while (len > 0) {
        bytes = read(fd, buf, len);
        if (bytes <= 0) {
            close(fd);
            return -1;
        }
        len -= bytes;
        buf += bytes;
    }
    close(fd);
    return 0;
}

static inline double xmin(double a, double b)
{
    return a < b ? a : b;
}

static int random_uniform(int min, int max)
{
    int val;
    int range;
    int copies;
    int limit;
    int ra;

    if (flb_random_bytes((unsigned char *) &val, sizeof(int))) {
        val = time(NULL);
    }
    srand(val);

    range  = max - min + 1;
    copies = (RAND_MAX / range);
    limit  = range * copies;
    ra     = -1;

    while (ra < 0 || ra >= limit) {
        ra = rand();
    }

    return ra / copies + min;
}

static int backoff_full_jitter(int base, int cap, int n)
{
    int temp;

    temp = xmin(cap, ipow(base * 2, n));
    return random_uniform(base, temp);
}

int main(int argc, char **argv)
{
    int tries;
    int seconds;
    for (tries = 0; tries < 10; tries++) 
    {
        seconds = backoff_full_jitter(FLB_SCHED_BASE, FLB_SCHED_CAP, tries);
        seconds = seconds + 1;
        printf("retry %2u - %4u backoff\n", tries, seconds);
    }
    return 0;
}
```

**Comment by donbowman** [comment]
I think we want the code more like this:

```
diff --git a/include/fluent-bit/flb_scheduler.h b/include/fluent-bit/flb_scheduler.h
index 457766f8..a6e43e48 100644
--- a/include/fluent-bit/flb_scheduler.h
+++ b/include/fluent-bit/flb_scheduler.h
@@ -27,8 +27,8 @@
 #include <fluent-bit/flb_output.h>
 
 /* Sched contstants */
-#define FLB_SCHED_CAP            2000
-#define FLB_SCHED_BASE           5
+#define FLB_SCHED_CAP            10
+#define FLB_SCHED_BASE           3
 #define FLB_SCHED_REQUEST_FRAME  10
 
 /* Timer types */
```

comments?

**Comment by donbowman** [comment]
i definitely recommend changing the code, as you can see it is not achieving the desired goals.

**Comment by tporeba** [comment]
I observed the same on kafka output, it would be nice to have configuration option to ease up on the backoff a bit as it escalates pretty quickly
```
[2021/07/12 13:00:23] [ warn] [engine] failed to flush chunk '32386-1626087613.285302602.flb', retry in 11 seconds: task_id=31, input=storage_backlog.2 > output=kafka.0 (out_id=0)
[2021/07/12 13:00:34] [ warn] [engine] failed to flush chunk '32386-1626087613.285302602.flb', retry in 16 seconds: task_id=31, input=storage_backlog.2 > output=kafka.0 (out_id=0)
[2021/07/12 13:00:51] [ warn] [engine] failed to flush chunk '32386-1626087613.285302602.flb', retry in 362 seconds: task_id=31, input=storage_backlog.2 > output=kafka.0 (out_id=0)
[2021/07/12 13:06:54] [ warn] [engine] failed to flush chunk '32386-1626087613.285302602.flb', retry in 1815 seconds: task_id=31, input=storage_backlog.2 > output=kafka.0 (out_id=0)
```
During the tests I ended up restarting the process instead of waiting as after requeueing from storage retry timeouts are back in lower range.

**Comment by JeffLuoo** [comment]
If running the fluent-bit in cluster, try remove the cpu and memory limit to see if that helps.

**Comment by donbowman** [comment]
the backoff algorithm is broken i think, i do not believe it is accurately reproducting the paper https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/

see my https://github.com/fluent/fluent-bit/issues/3597#issuecomment-857076504 comment.
If you compile that and run it, you will see the backoff in action. Its not related to the cpu or memory.

i don't think a 1hr backoff is ever good.

**Comment by JeffLuoo** [comment]
oh mb. I thought about the different issues. Increasing the CPU limit is not related to this issue.

**Comment by agup006** [comment]
If I understand this the solution is really to make max retry time configurable - @nokute78 Is this something we could add to Service configuration?

**Comment by donbowman** [comment]
Also, we need to fix the backoff. It has a math error of some type. I don't
see what, but the results are wrong.

**Comment by donbowman** [comment]
i know i'm not the only one facing this. I made a simple patch to my local tree, but i suspect others are quickly getting into 1hour pauses.

**Comment by nokute78** [comment]
I sent a patch https://github.com/fluent/fluent-bit/pull/4099 for now.

It is to make configurable a cap and base.

**Comment by edsiper** [comment]
thanks for the patch @nokute78, as soon CI finish we will merge it.

**Comment by edsiper** [comment]
FYI: just merged #4099

**Comment by agup006** [comment]
I know the bot has been auto closing this, however looks like this one has actually been merged. I’m closing for now
